File: ios/Muraka/Core/DesignSystem/StateViews.swift
import UIKit

/// The states every data-backed screen has to handle.
///
/// They live here rather than being rewritten per screen because the loading, empty and error
/// paths are where interfaces are usually thinnest — and in this app the error path is not an
/// edge case: a phone on a boat spends most of its life in it.
final class MessageStateView: UIStackView {
    private let titleLabel = UILabel()
    private let bodyLabel = UILabel()
    private let actionButton = UIButton(configuration: .filled())
    private var action: (() -> Void)?

    init(
        title: String,
        body: String,
        systemImage: String? = nil,
        actionTitle: String? = nil,
        action: (() -> Void)? = nil
    ) {
        super.init(frame: .zero)
        axis = .vertical
        alignment = .center
        spacing = 12
        isLayoutMarginsRelativeArrangement = true
        layoutMargins = UIEdgeInsets(top: 32, left: 32, bottom: 32, right: 32)

        if let systemImage {
            let imageView = UIImageView(image: UIImage(systemName: systemImage))
            imageView.tintColor = .secondaryLabel
            imageView.contentMode = .scaleAspectFit
            imageView.setContentHuggingPriority(.required, for: .vertical)
            addArrangedSubview(imageView)
        }

        titleLabel.text = title
        titleLabel.font = .preferredFont(forTextStyle: .headline)
        titleLabel.adjustsFontForContentSizeCategory = true
        titleLabel.textAlignment = .center
        titleLabel.numberOfLines = 0
        addArrangedSubview(titleLabel)

        bodyLabel.text = body
        bodyLabel.font = .preferredFont(forTextStyle: .body)
        bodyLabel.adjustsFontForContentSizeCategory = true
        bodyLabel.textColor = .secondaryLabel
        bodyLabel.textAlignment = .center
        bodyLabel.numberOfLines = 0
        addArrangedSubview(bodyLabel)

        // Always offer a way forward where one could conceivably help. A dead end with no
        // action is what makes a contributor close the app with a sighting still queued.
        if let actionTitle, let action {
            self.action = action
            actionButton.setTitle(actionTitle, for: .normal)
            actionButton.addTarget(self, action: #selector(runAction), for: .touchUpInside)
            addArrangedSubview(actionButton)
        }
    }

    @available(*, unavailable)
    required init(coder _: NSCoder) {
        // UIStackView's init(coder:) is non-failable, so this cannot return nil the way a
        // UIView subclass's can. Nothing in this app comes from a XIB, so reaching it is a
        // programmer error rather than a runtime condition.
        fatalError("Muraka builds its views in code; init(coder:) is unused")
    }

    @objc private func runAction() { action?() }
}

extension UIViewController {
    /// Shows a message the contributor can dismiss. Used for failures that are not a state
    /// of the screen — a retry that did not work, say.
    func presentMessage(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default))
        present(alert, animated: true)
    }
}

File: android/core/designsystem/src/main/kotlin/mv/muraka/core/designsystem/component/StateViews.kt
package mv.muraka.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

/**
 * The four states every data-backed screen has to handle.
 *
 * They live here rather than being rewritten per screen because the loading, empty and
 * error paths are where interfaces are usually thinnest - and in this app the error path
 * is not an edge case: a phone on a boat spends most of its life in it.
 */

@Composable
fun LoadingState(modifier: Modifier = Modifier, message: String = "Loading") {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp).semantics { contentDescription = message },
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        CircularProgressIndicator()
        Text(text = message, style = MaterialTheme.typography.bodyMedium)
    }
}

/**
 * Nothing here yet - and why, plus what to do about it.
 *
 * An empty state that only says "no sightings" wastes the one screen where the
 * contributor is most likely to need telling what the app is for.
 */
@Composable
fun EmptyState(
    title: String,
    body: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        icon?.let {
            Icon(
                imageVector = it,
                contentDescription = null,
                modifier = Modifier.size(40.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Text(text = title, style = MaterialTheme.typography.titleMedium, textAlign = TextAlign.Center)
        Text(
            text = body,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        if (actionLabel != null && onAction != null) {
            Button(onClick = onAction) { Text(actionLabel) }
        }
    }
}

/**
 * Something went wrong, with a way out.
 *
 * Always offers [onRetry] where a retry could conceivably help. A dead end with no action
 * is what makes a contributor close the app with a sighting still queued.
 */
@Composable
fun ErrorState(title: String, body: String, modifier: Modifier = Modifier, onRetry: (() -> Unit)? = null) {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(text = title, style = MaterialTheme.typography.titleMedium, textAlign = TextAlign.Center)
        Text(
            text = body,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        onRetry?.let { Button(onClick = it) { Text("Try again") } }
    }
}
